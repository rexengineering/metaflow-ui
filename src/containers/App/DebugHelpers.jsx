import React from "react";
import { Button,  makeStyles, Typography } from "@material-ui/core";
import WorkflowInstantiator from "../../components/WorkflowInstantiator";
import { fetchTasks } from "../../store/thunks/thunks";
import { useDispatch } from "react-redux";

const useStyles = makeStyles((theme) => ({
  root: {
    padding: theme.spacing(1),
  },
  deployment: {
    marginBottom: theme.spacing(2),
  },
  deploymentTitle: {
    marginBottom: theme.spacing(1),
  },
  button: {
    marginTop: theme.spacing(2),
    display: "block",
  },
}));

function DebugHelpers({ deployments, isAutomaticState, setIsAutomaticState, instanceID }) {
  const classes = useStyles();
  const isDeploymentsAvailable = Array.isArray(deployments) && !!deployments.length;
  const dispatch = useDispatch();

  return (
    <div className={classes.root}>
      <Typography variant="h5" align="center">
        Initiate Workflows
      </Typography>
      {!isDeploymentsAvailable && (
        <Typography>There are no workflows available</Typography>
      )}
      {Array.isArray(deployments) &&
        deployments.map((deploymentID) => (
          <div key={deploymentID} className={classes.deployment} data-testid="deployment">
            <Typography
              variant="body2"
              className={classes.deploymentTitle}
            >
              {deploymentID}
            </Typography>
            <WorkflowInstantiator deploymentID={deploymentID} instanceID={instanceID} />
          </div>
        ))}
      <Typography variant="h5" align="center">
        State Helpers
      </Typography>
      <Button
        color="secondary"
        type="button"
        variant="contained"
        onClick={() => dispatch(fetchTasks())}
        className={classes.button}
        data-testid="update-state-button"
      >
        Update state
      </Button>
      <Button
        variant="contained"
        type="button"
        onClick={(currentState) => setIsAutomaticState(!currentState)}
        className={classes.button}
        data-testid="toggle-state-button"
      >
        Toggle automatic state
      </Button>
      <Typography>
        Automatic state fetching is {isAutomaticState ? "On" : "Off"}
      </Typography>
    </div>
  );
}

export default DebugHelpers;
