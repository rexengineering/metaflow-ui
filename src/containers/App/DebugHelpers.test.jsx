import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import userEvent from "@testing-library/user-event";
import DebugHelpers from "./DebugHelpers";
import getStore from "../../rexui/store";

describe("<DebugHelpers />", () => {
  const wrappedDebugHelpers = (props, store = getStore({})) => (
    <Provider store={store}>
      <DebugHelpers {...props} />
    </Provider>
  );

  it("should render", () => {
    const { asFragment } = render(wrappedDebugHelpers());
    const initialRender = asFragment();

    expect(initialRender).not.toBeNull();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(wrappedDebugHelpers());
    const initialRender = asFragment();

    expect(initialRender).toMatchSnapshot();
  });

  it("should render elements for every deployment", () => {
    const deployments = ["deploymentId1", "deploymentId2", "deploymentId3"];
    render(wrappedDebugHelpers({ deployments }));
    const deploymentElements = screen.getAllByTestId("deployment");

    expect(deploymentElements).toHaveLength(deployments.length);
  });

  it("should dispatch fetchTasks when update state button is pressed", async () => {
    // const setIsAutomaticState = jest.fn()
    // const updateButton = screen.getByTestId("update-state-button");
    //
    // await waitFor(() => {
    //   userEvent.click(updateButton);
    // })
    // expect().toHaveBeenCalled();
  });

  it("should dispatch setIsAutomaticState when toggle state button is pressed", async () => {
    const setIsAutomaticState = jest.fn()
    render(wrappedDebugHelpers({ setIsAutomaticState }));
    const toggleButton = screen.getByTestId("toggle-state-button");

    await waitFor(() => {
      userEvent.click(toggleButton);
    })
    expect(setIsAutomaticState).toHaveBeenCalled();
  });
});
