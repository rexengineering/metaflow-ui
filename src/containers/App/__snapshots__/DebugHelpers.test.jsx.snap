// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DebugHelpers /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-5"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter"
    >
      Initiate Workflows
    </h5>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      There are no workflows available
    </p>
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter"
    >
      State Helpers
    </h5>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-8 MuiButton-containedSecondary"
      data-testid="update-state-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Update state
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-8"
      data-testid="toggle-state-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Toggle automatic state
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Automatic state fetching is Off
    </p>
  </div>
</DocumentFragment>
`;
